Add stm32cube framework

bjsowa opened the pull request:
I  tested it only on F4. Also I haven't tested the rollover feature, because I don't know how to do it without running the code for 50 days.

pablogs9 commented:
This looks good to me.

Could you provide some transport layer such as basic serial transport using STM HAL?

Also, could you integrate this into the CI? https://github.com/micro-ROS/micro_ros_platformio/blob/c9c410c9c7b93cca685cd19786978b0589f23ad9/.github/workflows/ci.yml#L17

pablogs9 commented:
@mergify backport humble foxy galactic

pablogs9 commented on platform_code/stm32cube/serial/micro_ros_transport.cpp:
```diff
@@ -82,19 +82,19 @@ size_t platformio_transport_read(struct uxrCustomTransport *transport,
         rhead <= rtail ? rtail - rhead : stream->rbuffer_size - rhead + rtail;
 
-    if (data_available >= len) break;
+    if (data_available >= len || ms_used >= timeout) break;
```
This should not wait until read data is >= len. This should wait until there is available 1 o more Bytes.

bjsowa replied:
fixed it

pablogs9 replied:
Cool, thanks

bjsowa commented:
> Also, could you integrate this into the CI?

I would integrate it with CI but I would need some help. For starters, I have a few questions:
1. Should I split the `ci` project into 2 projects (for example: `ci/arduino` and `ci/stm32cube`) to separate the code for different frameworks? 
2. Which [boards](https://docs.platformio.org/en/stable/frameworks/stm32cube.html#boards) should I choose for the CI?
3. Should the code be actually functional when flashed to the board or is it enough if it compiles? The stm32cube HAL requires a lot of boilerplate code for setting clocks, peripherals, etc. In my project, I use [Stm32CubeMX](https://www.st.com/en/development-tools/stm32cubemx.html) (through [stm32pio](https://github.com/ussserrr/stm32pio)) to genarate this code, but I'm not sure how it would work here.

pablogs9 commented:
> Should I split the ci project into 2 projects (for example: ci/arduino and ci/stm32cube) to separate the code for different frameworks?

Maybe you can add a target [here](https://github.com/micro-ROS/micro_ros_platformio/blob/main/ci/platformio.ini). But if you need to create another folder, it is ok for us.

> Which [boards](https://docs.platformio.org/en/stable/frameworks/stm32cube.html#boards) should I choose for the CI?

Maybe for initial support, we can add a board that you as a contributor can confirm that works. Just add the needed references in the README.md specifying that this has been tested in X hardware.

> Should the code be actually functional when flashed to the board or is it enough if it compiles? The stm32cube HAL requires a lot of boilerplate code for setting clocks, peripherals, etc. In my project, I use [Stm32CubeMX](https://www.st.com/en/development-tools/stm32cubemx.html) (through [stm32pio](https://github.com/ussserrr/stm32pio)) to generate this code, but I'm not sure how it would work here.

In general, if support for one platform/hardware is added, another user should be able to use it. If your contribution requires further steps to make a basic micro-ROS application works, feel free to add a section in the README.md explaining the dependencies and required steps for making in work.

bjsowa commented:
I added an example project and CI job. I didn't want to include generated code so I added a step in CI job that installs CubeMX and stm32pio which I use to generate the code before running platformio. 
The example project works on my custom board. Adding another boards should be fairly easy as it should be only a matter of creating a properly  configured `.ioc` project for the board. I don't have any other boards at hand unfortunately.
@pablogs9 any comments?

pablogs9 commented:
@Acuadros95 please review

Acuadros95 commented on platform_code/stm32cube/serial/micro_ros_transport.cpp:
```diff
@@ -0,0 +1,104 @@
+    rhead = rtail = thead = thead_next = ttail = 0;
+    HAL_UART_Receive_DMA(stream->uart, stream->rbuffer, stream->rbuffer_size);
+  }
+  open_count++;
```
Why use a integer for this?

Also, did you encounter any situation where open/close calls miss match?

bjsowa replied:
In my case, I used `rmw_uros_ping_agent` periodically to detect agent disconnects. The issue was that this function called open and close on the transport each time it was called, even when the transport was already opened. This is why I decided to ignore all nested open/close calls, hence I use integer to track the nesting level.

It's possible I used the library wrong. Let me know if that's the case or if you know a better way to deal with this problem.

Acuadros95 commented on ci/stm32cube/Src/main.c:
```diff
@@ -0,0 +1,104 @@
+  }
+}
+
+void HAL_UART_TxCpltCallback(UART_HandleTypeDef* huart) {
+  if (huart == &huart1) {
+    uart_transfer_complete_callback(&stream);
+  }
+}
```
Could we move this inside of the library?

bjsowa replied:
Implementing `HAL_UART_TxCpltCallback` inside the library would be problematic in case the user wants to use another UART port for something else. I don't see a clean way of doing this, any ideas?

Acuadros95 commented on .github/workflows/ci.yml:
```diff
@@ -1,30 +1,104 @@
+      fail-fast: false
+      matrix:
+        pio-environment:
+          - teensy41
```
```suggestion
```
Repeated

Acuadros95 reviewed (commented):
LGTM, will test it on monday.

Also, add the board to the [Supported boards](https://github.com/micro-ROS/micro_ros_platformio#supported-boards) list please.
